### PR TITLE
Feature/standard layout input

### LINKED ISSUE #365: [IMPROVEMENT] Input/output/update data structures as standard layout
## Background

The underlying [input, output and update data structures](https://power-grid-model.readthedocs.io/en/stable/user_manual/components.html) behind the data types retrieved by [initialize_array](https://power-grid-model.readthedocs.io/en/stable/api_reference/python-api-reference.html#power_grid_model.initialize_array) are generated using [code generation](https://github.com/PowerGridModel/power-grid-model/tree/main/code_generation).

The current implementation uses inheritance to define the relationship between components and the types they implement and are derived from. This keeps the data structures short and readable.

## Problem

The downside is that inheritance causes the data types are not guaranteed to obey standard layout. This may result in less efficient data structures, and also makes the code more bug-prone. As an example, #241 was caused by the fact that the data type does not have standard layout.

### Plain text with implicit reference operator conversion

```c++
// Example program
#include <iostream>
#include <limits>


using ID = int;

struct BaseInput { ID id; };
struct NodeInput{
    ID id;
    double u_rated;
    
    // Sonar Cloud warnings about `explicit` keyword and reinterpret_cast should be disabled here
    operator BaseInput&() { return reinterpret_cast<BaseInput&>(*this); }
    operator BaseInput const&() const { return reinterpret_cast<BaseInput const&>(*this); }

};

int main()
{
    NodeInput const node_input{ 5, 10.0 };
    BaseInput const& base = node_input;
    std::cout << base.id << '\n';
}
```

### Concepts

The standard layout solution can be achieved using a plain struct types without references to any base classes. For written code, that would cause a lot of code duplication, but since the code under consideration is generated, this is not an issue.

The notion of the "base" types are enabled by declaring them `concept`s, which can be tested using `static_assert`.

```cpp
template <typename T>
concept base_input_c = requires (T t) {
    { t.id } -> std::same_as<ID&>;
};

template <typename T>
concept node_input_c = base_input_c <T> && requires (T t) {
    { t.u_rated} -> std::same_as<double&>;
};

template <bool sym>
struct NodeInput {
    ID id{};
    double u_rated{};
};

static_assert(base_input_c<NodeInput<true>>);
static_assert(node_input_c<NodeInput<true>>);
static_assert(base_input_c<NodeInput<false>>);
static_assert(node_input_c<NodeInput<false>>);
```

Constructors of components will then take the following form (and equivalent for their update and output functions):

```cpp
class Base {
  public:
    explicit Base(base_input_c auto const& input_data): id_{input_data.id} {}

  private:
    ID id_{};
};

class Node: Base {
  public:
    explicit Node(node_input_c auto const& input_data): Base(input_data), u_rated_{input_data.u_rated} {}

  private:
    double u_rated_{};
};
```

This also will make the initializer list constructors a lot more readable for e.g. testing purposes.
Intellisense does provide suggestions on e.g. the member functions, but only when the type is `constraint auto obj`, not `constraint auto const& obj`.

### Composition

Another option could be to use [composition](https://en.wikipedia.org/wiki/Composition_over_inheritance), e.g. using data types like

```cpp
struct BaseOutput{
  ID id{};
};
struct DerivedOutput{
  BaseOutput base{};
  int other_field{};
};
```

but it becomes prohibitively non-trivial to query data fields of underlying types without getters and setters (e.g. we would need `transformer_output.line.base.id` to query the `id` field of a transformer).

### PR DESCRIPTION
**Pull request by mgovers**
Fixes #365 

Implements all input, output and update data structures as standard layout by using flat attribute expansion in the generated code. conversions to former base types is now done using `reinterpret_cast`

Relates to #241 and #276

### CODE REVIEW

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp [hunk 1]
```diff
@@ -21,36 +21,78 @@ struct BaseInput {
...
-struct NodeInput : BaseInput {
+static_assert(std::is_standard_layout_v<BaseInput>);
+
+struct NodeInput {
+    ID id;  // ID of the object
     double u_rated;  // rated line-line voltage
+
+    // implicit conversions to BaseInput
+    operator BaseInput&() { return reinterpret_cast<BaseInput&>(*this); }
+    operator BaseInput const&() const { return reinterpret_cast<BaseInput const&>(*this); }
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp:32]: every implicit conversion introduces 2 new sonar cloud code smells. that's 4 per base type, and 250 new ones in total.

I'm personally not a fan of adding `// NOSONAR` to every single line.

I've also read that it's also possible to do `*static_cast<BaseInput*>(static_cast<void*>(this))`. if the alignment is the same, it should actually be possible to do directly `*static_cast<BaseInput*>(this)`, but for some reason that's not working here. see also https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#1120 and https://godbolt.org/z/E884oTEqr

In addition, it also complains that we should use `std::bit_cast` when dealing with types that have no additional attributes (e.g. `SensorShortCircuitOutput`). In those cases, shouldn't we just use e.g. `using SensorShortCircuitOutput = BaseOutput` instead?

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp [hunk 2]
```diff
@@ -21,36 +21,78 @@ struct BaseInput {
...
 };
 
-struct BranchInput : BaseInput {
+static_assert(std::is_standard_layout_v<NodeInput>);
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp:35]: should i also check that the pointer to all base attribute members and/or alignment are identical? e.g.

```suggestion
static_assert(std::is_standard_layout_v<NodeInput>);
static_assert(std::alignment_of_v<NodeInput> == std::alignment_of_v<BaseInput>);
static_assert(offsetof(NodeInput, id) == offsetof(BaseInput, id));
```

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp (hunk 1 above)
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp:32]: Let's not make thing more difficult. We know what we are doing with `reintepret_cast` here. Just disable globally the warning from SonarCloud.

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp (hunk 2 above)
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp:35]: It should be check on `>=`. As long as the alignment of "child" class is bigger or equal than de "parent" class, it should be fine.

Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp:35]: assert of `offsetof` is good.

Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp:35]: @mgovers  is this comment resolved and the PR is ready for review again?

Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp:35]: no not yet, i've implemented, but i'm still struggling to find a good way to deal with the template case without getting 1000000 static checks per component type and it makes the files unreadable.

i can also add those static asserts to a separate file, e.g. `input_static_checks.hpp` that is then included in `update.hpp` so as to not affect the readability of the dataset declarations

any thoughts?

Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp]: put into separate directory `<...>/auxiliary/static_asserts/<...>.hpp`

**Comment by mgovers** [comment]
@TonyXiang8787 this is ready for review

**Review by mgovers — COMMENTED** [review]
should i also revert parts of https://github.com/PowerGridModel/power-grid-model/pull/276/files#diff-2bac4fa15db710784f8450b130e98864e112b2b5e4b104c0470f876d5894f4eeL138-R141 ? or is that not necessary? or maybe we can use `offsetof` also in those functions?

**Comment by mgovers** [comment]
@TonyXiang8787 did everything as discussed, including conversion to base types.

The following is created in #435 

> should i also revert parts of https://github.com/PowerGridModel/power-grid-model/pull/276/files#diff-2bac4fa15db710784f8450b130e98864e112b2b5e4b104c0470f876d5894f4eeL138-R141 ? or is that not necessary? or maybe we can use `offsetof` also in those functions?

**Review by TonyXiang8787 — APPROVED** [review]

**Pull request merged**